### PR TITLE
Feature/add hook doc and minor fix

### PR DESCRIPTION
**Pull request by romaninsh**
Adds passing of 'required' from hasOne to the field as well as document Hooks in great detail.

### CODE REVIEW

**Comment by DarkSide666** [comment]
why not adding docs of `required` too in this PR and that way fix https://github.com/atk4/data/issues/233 ? :)

File: docs/hooks.rst [hunk 1]
```diff
@@ -4,44 +4,177 @@
...
+:php:meth:`Model::loaded`:
+
+ - beforeInsert($m, &$data) (creating new records only)
+ - afterInsert($m, $id)
+ - beforeUpdate($m, &$data) (updating existing records only. Not executed if model is not dirty)
+ - afterUpdate($m)
+
+The $data argument will contain array of actual data (field=>value) to be saved, which you can
+use to withdraw certain fields from actually being saved into the database (by unsetting it's value).
+
```
Comment (DarkSide666) [inline docs/hooks.rst]: Tiny real-life example here would be good. I mean - when to use beforeSave and when to use beforeInsert/Update.

File: docs/hooks.rst [hunk 2]
```diff
@@ -4,44 +4,177 @@
...
+ - afterLoad($m). ($m will contain data). Break to unload and skip.
+
+For the deletion it's pretty similar:
+
+ - beforeDelete($m, $id). Unload and Break to preserve record.
+ - afterDelete($m, $id).
+
+A good place to clean-up delete related records would be inside afterDelete, although
+if your database consistency requires those related records to be cleaned up first, use
+beforeDelete instead.
```
Comment (DarkSide666) [inline docs/hooks.rst]: Maybe small example how to use beforeDelete for soft-delete functionality would be good here?

File: docs/hooks.rst [hunk 3]
```diff
@@ -4,44 +4,177 @@
...
+ - beforeUpdateQuery($m, $dsql_query)
+ - afterUpdateQuery($m, $statement). Executed before retrieving data.
+ - beforeInsertQUery($m, $dsql_query)
+ - afterInsertQuery($m, $statement). Executed before retrieving data.
+
+The delete has only "before" hook:
+
+ - beforeDeleteQuery($m, $dsql_query)
+
+Finally for queries there are several hooks:
```
Comment (DarkSide666) [inline docs/hooks.rst]: several? i see only one :)

File: docs/hooks.rst [hunk 4]
```diff
@@ -4,44 +4,177 @@
...
+    - "count"
+    - "update"
+    - "delete"
+    - "select"
+    - "field"
+    - "fx" or "fx0"
+
 
 
-More on hooks
```
Comment (DarkSide666) [inline docs/hooks.rst]: There are 3 more hooks in Agile Data:
```
- onlyFields
- normalize
- afterAdd
```

File: docs/hooks.rst (hunk 1 above)
Comment (romaninsh) [reply docs/hooks.rst]: added link to SoftDelete example here, which uses the technique.

File: docs/hooks.rst (hunk 2 above)
Comment (romaninsh) [reply docs/hooks.rst]: added link.

**Pull request merged**